src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
             return start < TextWindow.Position;
         }
 
+        // Allows underscores in integers, except at beginning and end
+        private void ScanNumericLiteralSingleInteger(StringBuilder builder, ref bool underscoreInWrongPlace, bool isHex, bool isBinary)
+        {
+            if (TextWindow.PeekChar() == '_')
+            {
+                underscoreInWrongPlace = true;
+            }
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
+            {
+                ch = TextWindow.PeekChar();
+                if (ch == '_')
+                {
+                    CheckFeatureAvailability(MessageID.IDS_FeatureDigitSeparator);
```

Comment:
This should probably be done once rather than on every `_`, as it appears in the VB implementation.